Don't load Apple/Google Pay for synced subscriptions that don't have payment due today and also require shipping

mattallan opened the pull request:
Fixes #3742

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->


When customers attempt to use Apple Pay or Google Pay to purchase a synchronized subscription product that has no upfront costs (i.e. trial period or delayed initial payment due to product being synced) and also requires shipping, they're met with an error on the page:

![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/f35a9c20-0a53-4d35-9f84-fa2f4a05760a)

Inside our `WC_Payments_Payment_Request_Button_Handler` class, we have code that prevents non-virtual (physical) products with free trials from being purchased with Apple/Google Pay, however there isn't any code to handle products that imitate a free trial period like what we do to handle synchronised subscription purchased outside of the synced date with no initial payment.

In this PR, I've introduced a new function to check if the product is a subscription product and has a free trial period (either trial set on the product or is a synced product), and requires shipping, and if so, prevent loading the Apple Pay/Google Pay buttons on the page.


#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Enable Apple Pay and Google Pay payment methods
2. Install and active Woo Subscriptions
3. Enable synchronised subscriptions in **WooCommerce > Settings > Subscriptions > Synchronise renewals**
![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/ff2a0b57-bcc5-492a-9433-952ae079a708)
4. Create a synced product that is not virtual and is not synced to today:
![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/a9f7164f-c652-4acf-ae31-cf1a1b4c8ee7)
5. While on the `develop` branch, view the product page and note the Payment Request buttons (PRB) are loaded. If you attempt to purchase this product with PRB, you will see the above error.
6. Check out this branch and confirm the PRB are no longer loaded on product pages for physical synced subscription products that

Other test cases to run:
 - Physical (non-virtual) variable synced subscriptions. The PRB shouldn't show on the product page if any variation isn't supported.
 - Virtual synced subscriptions. The PRB should show.
 - Physical synced products with a sync date set to today. The PRB should show
 - Physical synced subscription that is synced to another day but proration is enabled in **WC > Settings > Subscriptions**

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
Introduce new function to determine if product has trial period and needs shipping

Commit:
Add chagelog entry

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7582` or branch name `fix/3742-synced-subscriptions-payment-requests` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/3742-synced-subscriptions-payment-requests) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: d9cc5617ac1a21495041c1c897cfea260853388c
- Build time: 2024-02-06 05:40:08 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Fix undefined function in unit tests

Commit:
Fix psalm issues

james-allan commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -1700,4 +1700,44 @@ private function get_taxes( $product, $price ) {
+		}
+
+		return false;
+	}
```
@mattallan & I had a look at this and think there may be a less heavy handed approach to fix this issue. This function currently makes the whole variable product unpurchasable using express payment methods if 1 of the variations requires shipping. 

We agreed that the better way to go would be to use the `ajax_get_selected_product_data()` function or similar to limit the express payment method to eligible products. 

ie:

1. Change this `product_has_trial_and_needs_shipping()` function (and possibly its name) to check if **all** variations require shipping and have a free trial. 
2. Change `ajax_get_selected_product_data()` to throw an exception if the product is ineligible. 

That would have the effect of only loading the express payment method buttons if there is at least 1 eligible variation and hiding the express payment buttons if you choose a variation which is ineligible.

Mayisha replied:
Asking out of curiosity, do you have any type of product in mind that may have some variations that need shipping and some that do not need shipping? I did not think of this scenario initially :thinking:

james-allan replied:
It could be like a variable product where a customer can sign up to the virtual copy or a physical copy of a magazine. But because this logic includes shipping and free trial this could also apply to stores who offers a free trial option on a variable product. eg subscribe now or sign up to our free trial.

Mayisha reviewed (approved):
Thanks @mattallan. The fix is working as described :tada:

Mayisha commented:
@mattallan I have reviewed the changes and approved the PR as things are fixed. But you may need to consider the scenario @james-allan raised about variations.

elazzabi committed:
Always show Express payment buttons for Pay for Order page (#7535)

bborman22 committed:
Check for store api rest route in parameter (#7558)

frosso committed:
chore: cleanup unused hook dependency (#7609)

mgascam committed:
Avoid returning Stripe error message when capturing more than order amount (#7581)

LCmry committed:
Do not show WooPay button on external/affiliate product page (#7589)

Commit:
Refactor product_has_trial_and_needs_shipping() to only return true if ALL variations pass

Commit:
Whenever a variation is selected, check if it's valid to purchase with express payment methods

Commit:
Add unit tests to product_has_trial_and_needs_shipping()

mattallan commented:
Thanks @Mayisha and @james-allan. I've refactored the `product_has_trial_and_needs_shipping()` function so that if this function is passed a variable subscription, it will only return true if all the variations have a trial and require shipping.

This means if you have a subscription that has one variation that is valid for express payment methods and another that is not, the button will still appear on the product page for it and only show for variations that are supported, here's a quick screencast showcasing this:

![Screen Capture on 2023-11-03 at 12-48-54](https://github.com/Automattic/woocommerce-payments/assets/2275145/31b78948-b92d-48bb-8030-ce2ec94d7b22)

james-allan committed:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

james-allan committed:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

james-allan commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -1703,4 +1707,46 @@ private function get_taxes( $product, $price ) {
+
+			// If product is synced, check if the first payment is upfront or today (i.e. no trial period). If product is not synced, check if it has a trial period.
+			if ( WC_Subscriptions_Synchroniser::is_product_synced( $product ) ) {
+				if ( WC_Subscriptions_Synchroniser::is_payment_upfront( $product ) || WC_Subscriptions_Synchroniser::is_today( WC_Subscriptions_Synchroniser::calculate_first_payment_date( $product, 'timestamp' ) ) ) {
```
```suggestion
				if ( WC_Subscriptions_Synchroniser::is_payment_upfront( $product ) ) {
```

You can get rid of the `WC_Subscriptions_Synchroniser::is_today()` check because the `is_payment_upfront()` function checks if the product is synced to today [here](https://github.com/Automattic/woocommerce-subscriptions-core/blob/5.5.0/includes/class-wc-subscriptions-synchroniser.php#L548-L549).

mattallan replied:
Thanks @james-allan. Looks like `is_payment_upfront` does infact also call `calculate_first_payment_date` and compares the `'ymd'` of this date and current time to check if it's the same ✋

james-allan reviewed (changes requested):
Hey @mattallan.

Changes look good and works. I left a comment suggesting that we can remove the `is_today()` check. I believe that isn't necessary given the `is_payment_upfront()` function includes an equivalent check. :)

Commit:
Update includes/class-wc-payments-payment-request-button-handler.php

Co-authored-by: James Allan <james.allan@automattic.com>

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Fix unit tests now that is_today isn't being used

Commit:
fix psalm issues

mattallan commented:
Thanks for the suggestion @james-allan. I had to fix the psalm issues and unit tests but this is good for a final review

Edit: I've ran through the standard tests on the latest changes

james-allan committed:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Rename function to is_invalid_subscription_product()

Commit:
Don't show PRB for virtual synced subscription products

Commit:
When a variation is pre-selected on page load using GET params, check if it's valid

Commit:
After adding a variation to the cart that doesn't support PRB, don't show PRB on the page load after

mattallan commented:
While copying these changes over to our Stripe gateway, I found a few more issues over there that are also issues in WooPayments.

Here are the individual fixes I've pushed up to this PR:
- [x] Don't display the PRB on the product page for virtual synced subscriptions that have no upfront payment. This is to prevent incorrect totals that don't account for things like a trial period or a prorated initial payment. Fixed by de80fb0fc
- [x] If a variation is already selected on page load (i.e. using `/?attribute_{name}=value` in URL), don't show the PRBs on page load if that variation isn't supported. Fixed by b9d2ba59c
- [x] After adding a variation to the cart that doesn't support PRBs, don't show PRBs on the page load that occurs while the unsupported variation is still selected. Fixed by 47374a0b4

@james-allan did you mind taking this PR for another spin? 🙏 Thanks!

Commit:
Add changelog entries for the extra fixes

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Fix unit tests

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Use updated function name in PHPUnit test function name

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

Commit:
Reword the changelog entry

Commit:
Merge branch 'develop' into fix/3742-synced-subscriptions-payment-requests

dmvrtx commented:
Closing, ECE has changed much since then the PR was open and has a better support of subscription products of different types and options.
